### ISSUE TITLE
Increase recommended default RetryPeriod for LeaderElection

### ISSUE THREAD

**Issue by linxiulei** [kubernetes/kubernetes#125861]
### What would you like to be added?

Review the default value of RetryPeriod for LeaderElection from current 2s to probably 5s or some values close to that.

### Why is this needed?

According to source [code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/config/v1alpha1/defaults.go#L37-L45)

	if obj.LeaseDuration == zero {
		obj.LeaseDuration = metav1.Duration{Duration: 15 * time.Second}
	}
	if obj.RenewDeadline == zero {
		obj.RenewDeadline = metav1.Duration{Duration: 10 * time.Second}
	}
	if obj.RetryPeriod == zero {
		obj.RetryPeriod = metav1.Duration{Duration: 2 * time.Second}
	}

LeaseDuration is 15s, RenewDeadline 10s however RetryPeriod is 2s, which is way lower than former two and causing visible amount of leader election renewal requests to apiserver, which could be excessive given LeaseDuration is 15s long.

So maybe increasing it to 1/3 of LeaseDuration still gives it sufficient headroom for retry.

**Comment by linxiulei** [comment]
Alternatively, we can also have dynamical retry period values that decreases when there are failures in previous attempts

**Comment by omerap12** [comment]
Hey, can I work on this issue? @cici37  @linxiulei

**Comment by omerap12** [comment]
I'll assign this to myself for now. If there are any objections, I'll remove my assignment.
/assign

**Comment by aojea** [comment]
before anything can we do some numbers to understand the implications of each number in terms of api requests and when those requests will happen? 

retryPeriod is using to acquire the lease

https://github.com/kubernetes/kubernetes/blob/a2106b5f73fe9352f7bc0520788855d57fc301e1/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L249-L270

and to renew it 

https://github.com/kubernetes/kubernetes/blob/a2106b5f73fe9352f7bc0520788855d57fc301e1/staging/src/k8s.io/client-go/tools/leaderelection/leaderelection.go#L279-L300

**Comment by linxiulei** [comment]
@aojea the description matched my observation -- the leader election request is sent every 2s by default. For the complete understanding of those values (pls correct me if I'm wrong):

__RetryPeriod__: the interval between LE requests (acquiring or renewing)
__RenewDeadline__: the timeout for LE requests (acquiring or renewing)
__LeaseDuration__: the duration of a lease to be valid

**Comment by aojea** [comment]
Yeah, I'm saying that this seems can impact the time to react on a leader change, or not? 
and  if everything works fine at first try it should have zero impact the retry period, no?

**Comment by linxiulei** [comment]
I see, the impact could be increasing the latency of a non-leader becoming a leader when the previous leader stops renewing. But it only impacts non-leaders so it should be safe to bump the time for leaders. 

We can optimize non-leaders with this [PR](https://github.com/kubernetes/kubernetes/pull/122049). With that we can even decrease the latency without adding more calls to apiserver.

**Comment by Jefftree** [comment]
I've been working on parts of this codebase and want to make sure our understandings are aligned. The goal is to reduce total reads and writes to the server for lease renewal. Raising this timeout will reduce requests to apiserver while slightly increasing the time to recover on a leader change? One thing to note here is that `config.ReleaseOnCancel` is usually not set so leases usually terminate via expiry. Worst case transition time is already 15s (leaseDuration) + 2s (retryPeriod) = 17s so changing that period to 15+5=20s is proportionally not as big of a difference.

**BEFORE:** 
In steady state with one leader, 3 candidates total:
- leader renews every 2 seconds under fast path (1 write/2s)
- 2 remaining candidates go through the tryAcquireOrRenew path, read that the lease is held and unexpired, and repeat the loop every 2s. (2 read/2s)

If lease is expired (up to 15s):
- other candidates will acquire within 2s

**AFTER:**
- leader renews every 5 seconds under fast path (1 write/5s)
- 2 remaining candidates read every 5s, (2 read/5s)

If lease is expired (up to 15s):
- other candidates will acquire within 5s

**Comment by linxiulei** [comment]
Friendly ping @jpbetz @liggitt @deads2k @wojtek-t

**Comment by liggitt** [comment]
> * 2 remaining candidates read every 5s, (2 read/5s)

is there a way to slow the write by the leader without more than doubling the recovery time by the non-leaders?

**Comment by linxiulei** [comment]
> > * 2 remaining candidates read every 5s, (2 read/5s)
> 
> is there a way to slow the write by the leader without more than doubling the recovery time by the non-leaders?

I believe https://github.com/kubernetes/kubernetes/pull/122049 is a way to go. If that PR is merged, the non-leaders can theoretically have the recovery time as low as the watch latency permits although there are other minor risks in discussion of that PR.
